{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true,
    "toc": "true"
   },
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev2\"><a href=\"#Description-of-CROP\"><span class=\"toc-item-num\">0.1&nbsp;&nbsp;</span>Description of CROP</a></div><div class=\"lev1\"><a href=\"#Objective-Function\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Objective Function</a></div><div class=\"lev2\"><a href=\"#Probability-of-biochemical-feasibility\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Probability of biochemical feasibility</a></div><div class=\"lev2\"><a href=\"#Probability-of-thermodynamic-feasibility\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Probability of thermodynamic feasibility</a></div><div class=\"lev2\"><a href=\"#Maximum-a-posteriori-(MAP)-estimation\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Maximum <em>a posteriori</em> (MAP) estimation</a></div><div class=\"lev1\"><a href=\"#Resolving-incorrect-predictions-by-restoring-growth\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Resolving incorrect predictions by restoring growth</a></div><div class=\"lev1\"><a href=\"#Resolving-incorrect-predictions-by-suppressing-growth\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Resolving incorrect predictions by suppressing growth</a></div><div class=\"lev2\"><a href=\"#Generic-solution-strategy\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Generic solution strategy</a></div><div class=\"lev3\"><a href=\"#Resolving-the-nonlinearity-in-the-general-solution-strategy\"><span class=\"toc-item-num\">3.1.1&nbsp;&nbsp;</span>Resolving the nonlinearity in the general solution strategy</a></div><div class=\"lev2\"><a href=\"#GrowMatch\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>GrowMatch</a></div><div class=\"lev2\"><a href=\"#ModelSEED\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>ModelSEED</a></div><div class=\"lev2\"><a href=\"#CROP\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>CROP</a></div><div class=\"lev3\"><a href=\"#ABC-model\"><span class=\"toc-item-num\">3.4.1&nbsp;&nbsp;</span>ABC model</a></div><div class=\"lev3\"><a href=\"#Robustness\"><span class=\"toc-item-num\">3.4.2&nbsp;&nbsp;</span>Robustness</a></div><div class=\"lev4\"><a href=\"#LP-dung-FBA-implementation\"><span class=\"toc-item-num\">3.4.2.1&nbsp;&nbsp;</span>LP dung-FBA implementation</a></div><div class=\"lev4\"><a href=\"#Dual-of-LP-dung-FBA-Implementation\"><span class=\"toc-item-num\">3.4.2.2&nbsp;&nbsp;</span>Dual of LP dung-FBA Implementation</a></div><div class=\"lev4\"><a href=\"#MILP-dung-FBA-Implementation\"><span class=\"toc-item-num\">3.4.2.3&nbsp;&nbsp;</span>MILP dung-FBA Implementation</a></div><div class=\"lev4\"><a href=\"#Dual-of-MILP-dung-FBA--Implementation\"><span class=\"toc-item-num\">3.4.2.4&nbsp;&nbsp;</span>Dual of MILP dung-FBA  Implementation</a></div><div class=\"lev3\"><a href=\"#Size\"><span class=\"toc-item-num\">3.4.3&nbsp;&nbsp;</span>Size</a></div><div class=\"lev4\"><a href=\"#Reduced-bi-level-MILP-dung-FBA-implementation\"><span class=\"toc-item-num\">3.4.3.1&nbsp;&nbsp;</span>Reduced bi-level MILP dung-FBA implementation</a></div><div class=\"lev4\"><a href=\"#Partially-reduced-bi-level-MILP-dung-FBA-implementation\"><span class=\"toc-item-num\">3.4.3.2&nbsp;&nbsp;</span>Partially reduced bi-level MILP dung-FBA implementation</a></div><div class=\"lev4\"><a href=\"#Biomass-substituted-partially-reduced-bi-level-MILP-dung-FBA-implementation\"><span class=\"toc-item-num\">3.4.3.3&nbsp;&nbsp;</span>Biomass-substituted partially reduced bi-level MILP dung-FBA implementation</a></div><div class=\"lev3\"><a href=\"#Handling-multiple-conditions-simultaneously\"><span class=\"toc-item-num\">3.4.4&nbsp;&nbsp;</span>Handling multiple conditions simultaneously</a></div><div class=\"lev4\"><a href=\"#Multiple-condition-biomass-substituted-partially-reduced-bi-level-MILP-dung-FBA-Implementation\"><span class=\"toc-item-num\">3.4.4.1&nbsp;&nbsp;</span>Multiple condition biomass-substituted partially reduced bi-level MILP dung-FBA Implementation</a></div><div class=\"lev4\"><a href=\"#Multiple-condition-nutrient-limited-partially-reduced-bi-level-MILP-dung-FBA-Implementation\"><span class=\"toc-item-num\">3.4.4.2&nbsp;&nbsp;</span>Multiple condition nutrient-limited partially reduced bi-level MILP dung-FBA Implementation</a></div><div class=\"lev1\"><a href=\"#References\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>References</a></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Description of CROP ##\n",
    "\n",
    "Here, we describe the [Consistent Reproduction Of growth/no-growth Phenotype (CROP)](http://journals.plos.org/ploscompbiol/article/file?type=supplementary&id=info:doi/10.1371/journal.pcbi.1003126.s011) algorithm of FARM. This optimization-based algorithm reconciles experimentally observed growth phenotypes with computationally predicted growth phenotypes by suggesting reaction additions/deletions to the metabolic network.\n",
    "Our exposition is split into three sections.\n",
    "\n",
    "* We describe the novel objective function that CROP uses to incorporate diverse evidence. \n",
    "\n",
    "We then describe our complete solution in the two situations where CROP is needed: \n",
    "* when no-growth is computationally predicted, but growth is experimentally observed; and \n",
    "* when growth is computationally predicted, but no-growth is experimentally observed.\n",
    "\n",
    "In the last section, we also provide a reference implementation that can be executed within this notebook for each variation of the CROP algorithm .\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Objective Function\n",
    "\n",
    "CROP maximizes an objective function that has weights associated with each reaction. These reactions are represented as being irreversible, where reversible reactions are split into two irreversible reactions, so that CROP can also choose among the reaction directions. \n",
    "\n",
    "We derive the weights using maximum *a posteriori* (MAP) estimation. This estimation requires the probabilities that a reaction is biochemically and thermodynamically feasible, because CROP’s purpose is to identify a stoichiometric matrix that defines the feasible space of reaction fluxes.\n",
    "\n",
    "## Probability of biochemical feasibilit\n",
    "\n",
    "A reaction is biochemically feasible if the net flux is positive in the direction of the reaction. Spontaneous reactions are always biochemically feasible. Transport reactions that require a transporter are biochemically feasible if and only if the genome encodes the transport function. Non-spontaneous enzyme-catalyzed reactions are biochemically feasible if and only if the genome encodes the enzyme function. The probability that a non-spontaneous reaction is biochemically feasible depends on the probability that the genome encodes the required function.\n",
    "Transport reactions were assigned a probability of one if they had associated evidence in NeurosporaCyc; otherwise they were assigned a probability of zero. Non-spontaneous enzyme-catalyzed reactions were assigned a probability from the enzyme function predictor EFICAz [[1]](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-107). When there was additional evidence available in NeurosporaCyc, such as genome annotation, pathway prediction, and experimental evidence from the literature, this evidence was used to update the assigned probabilities.\n",
    "\n",
    "## Probability of thermodynamic feasibility \n",
    "\n",
    "A reaction (in a given direction) is thermodynamically feasible if and only if its Gibbs free energy $\\Delta G' < 0$. $\\Delta G'$  is a function of the Gibbs free energy at standard state $\\Delta G^0$ and of the substrate/product concentrations. This function is $\\Delta G'=\\Delta G^0 + RT\\ln(Q)$, where $R$ is the gas constant, $T$ is the temperature in degrees Kelvin, and $Q$ is the mass-action ratio. For a biochemical equation of the form $aA + bB \\rightarrow cC + dD$, we have:\n",
    "\n",
    "$$\\begin{array}{rcl}\n",
    "    \\Delta G' & = & \\Delta G^0+RT\\ln\\left (\\frac{[C]^c[D]^d}{[A]^a[B]^b}\\right)\n",
    "\\end{array}$$\n",
    "    \n",
    "We can estimate $\\Delta G'$  from estimates of $\\Delta G^0$ [[2]](https://www.ncbi.nlm.nih.gov/pubmed/18645197) and from physiological ranges of metabolite concentrations. We can further assess the variance of $\\Delta G'$  from the variance of $\\Delta G^0$. Taken together, we can use the approach of Fleming et al. [[3]](https://www.ncbi.nlm.nih.gov/pubmed/19783351) to estimate the probability that a given reaction direction is thermodynamically feasible.\n",
    "Yet, when using this approach, we found that thermodynamically infeasible reactions were not sufficiently penalized, so we developed a novel way to estimate the probability that a given reaction direction is thermodynamically feasible. The reason that thermodynamically infeasible reactions were not sufficiently penalized was because metabolite concentrations were allowed to freely vary within the physiological range. We could not address this problem directly, because we do not have comprehensive measurements of metabolite concentrations across conditions. We do not even have comprehensive measurements of concentrations of cofactors, such as ATP, across conditions. This situation applies to all organisms, not only Neurospora [[4]](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002166). \n",
    "However, metabolite concentrations enter into equation (1) as ratios, not as absolutes. And these ratios -- such as $\\frac{NAD^+}{NADH}$, $\\frac{NADP^+}{NADPH}$, and $\\frac{ATP}{ADP}$ -- are bounded above and below by thermodynamics [[5]](https://www.ncbi.nlm.nih.gov/pubmed/16299075). Thus, we used known bounds on the ratios of currency metabolites to bring our thermodynamic probability estimates more in line with known thermodynamic feasibility. To illustrate, consider the reaction:\n",
    "\n",
    "$$glyoxylate + NADPH + H+ \\rightarrow glycolate + NADP+$$\n",
    "\n",
    "with $\\Delta G^0$  estimate $-7.68 \\frac{kcal}{mol}$. We have: \n",
    " \n",
    "$$\\Delta G' = -7.68 + RT\\ln\\left(\\frac{[glyoxylate][NADPH][H^+]}{[glycolate][NADP^+]}\\right)$$\n",
    "\n",
    "The maximum thermodynamic ratio of $\\frac{[NADP^+]}{[NADPH]}$ has been estimated to be 1.1 [[2]](https://www.ncbi.nlm.nih.gov/pubmed/18645197), and it is thought that organisms try to maintain this ratio near its maximum. Thus, we assumed that in the cytosol $\\frac{[NADP+]}{[NADPH]}=1.1$. To estimate $\\Delta G'$,  we also assumed a neutral pH (i.e. $[H^+]=10^{-7}$).\n",
    "\n",
    "$$\\begin{array}{rcl}\n",
    "\\Delta G' & = & -7.68 + RT\\ln\\left(\\frac{10^{-7}[glyoxylate]}{1.1[glycolate]}\\right) \\\\\n",
    "& = & -7.68 + 16RT\\ln\\left(\\frac{[glyoxylate]}{[glycolate]}\\right)\\\\\n",
    "\\end{array}$$\n",
    " \n",
    "Now, we can use the physiological range of metabolite concentrations to constrain $[glyoxylate]$ and $[glycolate]$. But since we’ve already used the maximum thermodynamic ratio of $\\frac{[NADP^+]}{[NADPH]}$, our confidence interval for   is much tighter, so the probability estimate is much more reliable. To our knowledge, using thermodynamic bounds on metabolite concentration ratios to obtain more accurate probabilistic estimates of thermodynamic feasibility is novel.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Maximum *a posteriori* (MAP) estimation ##\n",
    "Our derivation of weights associated with each reaction is motivated by MAP estimation. The reactions are irreversible reactions from a metabolic database, such as MetaCyc. We begin by defining a binary variable $Z_i$ for each reaction $i$ that denotes whether it is physiologically feasible for Neurospora to carry flux through reaction $i$ in a condition:\n",
    "\n",
    "$$Z_i = \\left\\{\n",
    "\\begin{array}{cl}\n",
    "1 & \\text{reaction $i$ is physiologically feasible} \\\\\n",
    "0 & \\text{otherwise} \\\\\n",
    "\\end{array}\\\\ \\right.$$\n",
    "\n",
    "$Z=1$ corresponds to the event that reaction $i$ belongs in the \"true\" metabolic network. Thus, CROP’s purpose is to decide which $Z_i$’s should be one and which should be zero; this decision specifies the metabolic network.\n",
    "We relate our data on probabilities of biochemical and thermodynamic feasibility (described above) to probabilities of physiological feasibility. \n",
    "\n",
    "$$\\begin{array}{rcl}\n",
    "p_i & = & P(\\text{reaction $i$ is physiologically feasible}) \\\\\n",
    "& = & P(\\text{reaction $i$ is biochemically and thermodynamically feasible}) \\\\\n",
    "& = & P(\\text{reaction $i$ is biochemically feasible})P(\\text{reaction $i$ is thermodynamically feasible})\n",
    "\\end{array}$$\n",
    "\n",
    "We then use our data $p_i$ to formulate the MAP estimate for the vector of $Z_i$’s. This vector estimate is denoted by  $\\hat{Z}$. \n",
    "$$\\begin{array}{rcl}\n",
    "\\hat{Z} & = & \\underset{Z_i}{\\text{argmax}}\\prod_i P(Z_i|p_i) \\\\ \n",
    "& = & \\underset{Z_i}{\\text{argmax}}\\sum_i \\ln\\left[P(Z_i|p_i)\\right]  \\\\\n",
    "\\end{array}$$\n",
    "\n",
    "The last equality holds true because the logarithm is a monotonically increasing function, so the vector of $Z_i$'s that maximize $P(Z_i|p_i)$ also maximize $\\ln\\left[P(Z_i|p_i)\\right]$.\n",
    "By definition of $p_i$, we know that:\n",
    "\n",
    "$$\\begin{equation}P(Z_i=1\\ |\\ p_i) = p_i\\end{equation}$$\n",
    "and similarly,\n",
    "$$\\begin{equation}P(Z_i=0\\ |\\ p_i) = 1-p_i\\end{equation}$$\n",
    "\n",
    "Now, we represent  $P(Z_i | p_i)$ as a mixture model, which “mixes” the cases where $Z_i=1$ and $Z_i=0$:\n",
    "$$\\begin{equation}\n",
    "\\begin{array}{rcl}\n",
    "P(Z_i|p_i) & = & Z_iP(Z_i=1\\ |p_i)+(1-Z_i)P(Z_i=0\\ |p_i) \\\\\n",
    "& = & Z_ip_i + (1-Z_i)(1-p_i)\\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "We log-transform $(6)$ to obtain a form that matches $(3)$.\n",
    "\n",
    "$$\\begin{equation}\\ln\\left[P(Z_i|p_i)\\right] = \\ln\\left[Z_ip_i + (1-Z_i)(1-p_i)\\right]\\end{equation}$$\n",
    "\n",
    "We would like to optimize  equation $(7)$, as shown in equation $(3)$, subject to the constraints of the network. (The full optimization is discussed in the sections below.) Because the $Z_i$ variables are binary, this optimization problem is very computationally demanding – in fact, it is NP-hard. So that this already demanding optimization problem is tractable using available solvers, its objective function must be linear or quadratic, and for speed, it is desirable for it to be linear.\n",
    "So we sought a linear approximation to equation $(11)$. To model this logarithmic function as a linear one, we appeal to the logarithm’s first-order Taylor series approximation around 0: $\\ln(1+x)≈x$ for $-1<x\\leq 1$. The appeal is implemented as:\n",
    "\n",
    "$$\\begin{equation}\\begin{array}{rcl}\n",
    "\\ln\\left[Z_ip_i+(1-Z_i)(1-p_i)\\right] \n",
    "& = & \\ln\\left[Z_ip_i + (1 - Z_i-p_i + Z_ip_i)\\right] \\\\\n",
    "& = & \\ln\\left[ 1 + ( 2Z_ip_i - Z_i-p_i)\\right] \\\\\n",
    "& = & 2Z_ip_i -Z_i -p_i \\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "To exclude the possibility that $Z_ip_i+(1-Z_i)(1-p_i)=0$, which would pose a problem for the logarithm,\n",
    "we only need to assume that $0<p_i<1$.\n",
    "\n",
    "We now put the above steps together to show our linear approximation for the MAP estimate\n",
    "$$\\begin{equation}\\begin{array}{rcl}\n",
    "\\hat{Z} & = & \\underset{Z}{\\text{argmax}}\\sum_i\\ln\\left[P(Z_i|p_i)\\right] \\\\\n",
    "& \\approx & \\underset{Z}{\\text{argmax}}\\sum_i2Z_ip_i-Z_i - p_i \\\\\n",
    "& = & \\underset{Z}{\\text{argmax}}\\sum_i2Z_ip_i-Z_i \\\\\n",
    "& = & \\underset{Z}{\\text{argmax}}\\sum_iZ_i(2p_i-1) \\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "where terms without $Z_i$ were removed because they are not involved in the optimization. We simplify by multiplying through by $\\frac{1}{2}$,\n",
    "$$\\hat{Z}  \\approx \\underset{Z}{\\text{argmax}}\\sum_iZ_i(p_i-0.5)$$\n",
    "\n",
    "which can be stated in matrix notation as:\n",
    "$$\\begin{equation}\\hat{Z} \\approx \\underset{Z}{\\text{argmax }} Z(p-0.5)\\end{equation}$$\n",
    "This implies that our objective function is: $Z(p-0.5)$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Resolving incorrect predictions by restoring growth #\n",
    "In this section, we mathematically deal with conditions where no-growth is predicted, but growth is experimentally observed. (A condition is defined by what genes are knocked-out and what nutrients are available.) The goal here is to identify model changes that restore growth in a given condition.\n",
    "There are several model changes that can restore growth. One model change that can restore growth is to assign an isozyme to the reaction that is catalyzed by the knocked-out gene, so that the gene knockout does not lead to reaction removals. This change is implemented by modifying the gene-protein-reaction (GPR) associations. Another model change that can restore growth is to reduce the number of essential compounds required for growth. This is implemented by changing the biomass composition in the stoichiometric matrix. If neither of these model changes are appropriate, we can restore growth by adding reactions that are not in the model from a universal pathway database such as MetaCyc. To identify the reactions to add to the model to restore growth, we implemented a mixed-integer linear program (MILP). Consider a stoichiometric matrix $S$ with irreversible reactions from a universal pathway database that includes reactions from the model, a vector of binary variables $z$ with associated non-negative weights corresponding to reactions in the database, and a vector of reaction upper bounds  $U$. For example, if the condition limits the glucose uptake reaction to $10 \\frac{mmol}{g\\cdot Dry\\ weight\\cdot hour}$, then we set $U_{glucose}=10$, and for all other reactions we set $U=1000$, as a loose upper bound. If the condition includes a gene knockout, then for reactions $KO$ that require the gene, we set $U_{KO}=0$. Now, assume that the ATP flux must be at least `atp_maintenance`, and that growth, represented by $v_{biomass}$, must be at least `minimal_growth`. Then, the general form of the MILP is:\n",
    "\n",
    "$$\n",
    "\\begin{array}{lll}\n",
    "\\underset{z}{\\min} & weights^Tz  & \\\\\n",
    "\\text{subject to:} & Sv = 0 & \\\\\n",
    "                   & 0 \\leq v_i\\leq U_i &  \\text{for } i\\in \\text{ model} \\\\\n",
    "                   & 0 \\leq v_i\\leq U_i\\cdot z_i & \\text{for } i\\in \\text{ database} \\\\\n",
    "                   & v_{ATP} \\geq \\text{ atp maintenance }  & \\\\\n",
    "                   & v_{biomass} \\geq \\text{minimal growth} & \\\\\n",
    "                   & z\\in\\{0,1\\} & \\\\\n",
    "\\end{array}\n",
    "$$\n",
    "\n",
    "\n",
    "The main difference between MILPs is how they assign weights for reactions from the database. GrowMatch [[6]](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000308) treats all reactions as having the same weight, so it minimizes the total number of reactions to add. The Model SEED [[7](https://www.ncbi.nlm.nih.gov/pubmed/19555510),[8](http://www.nature.com/nbt/journal/v28/n9/full/nbt.1672.html)] assigns these weights based on diverse evidence. Since there is one binary variable per reaction in the database, these MILPs can be very large, so they are time-consuming to solve.\n",
    "\n",
    "Our formulation is similar to program $(11)$. Since program $(11)$ is a minimization, the weights in our objective function (derived in the previous section) should be $0.5-p$.  However, this means that if the probability of a reaction is greater than $50\\%$, the solution set will include that reaction, even if it is not strictly required for growth. To ensure that unnecessary reactions are not added to the model, we require that our weights be non-negative. So, we slightly modify our objective function, and instead use: $weights=1-p$ , where $0<p<1$. \n",
    "\n",
    "Rather than solve the full MILP, we can first solve a linear relaxation of the MILP. This linear relaxation allows all of the variables $z_i$, which were formerly restricted to be binary, to lie anywhere between 0 and 1, inclusive.  \n",
    "\n",
    "The idea behind this linear relaxation is that it minimizes a weighted sum of flux through the database reactions. Thus, reactions not assigned flux in the relaxation are not needed to restore growth; conversely, reactions assigned flux are likely to have a low weight and contribute to restoring growth. Then, the set of reactions whose $z_i>0$ in the linear relaxation becomes the set of reactions to add to the model, and these reactions restore growth. \n",
    "The set of reactions returned by the relaxation--termed the *relaxation set*-- is not necessarily optimal for $(11)$. However, we found this set of reactions to be very similar to the optimal set in practice, and furthermore we could efficiently improve the set in two ways. \n",
    "\n",
    "1. We can treat the *relaxation set* as if it were the set of all database reactions, and then optimize program $(11)$ normally. Because the relaxation set is much smaller than the set of all database reactions, this optimization runs remarkably quickly. \n",
    "2. We can solve program $(11)$ using the *relaxation set* as a high-quality feasible starting point (often called a warm-start), which can save the optimizer substantial computational time [[9]](http://link.springer.com/article/10.1023/A:1011416130639).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Resolving incorrect predictions by suppressing growth #\n",
    "In this section, we mathematically deal with conditions where growth is predicted, but no growth is experimentally observed. The goal here is to identify model changes that suppress *in silico* growth in a given condition. Importantly, these model changes should not have the side-effect of suppressing growth in other conditions where growth is experimentally observed, such as the condition of wild-type in minimal medium.\n",
    "There are several model changes that can suppress growth. These are basically the reverse of those that restore growth. If the knocked-out gene encodes an isozyme, then removing its isozyme, or requiring both gene products as subunits of an enzyme complex, will lead to reaction removal. Reaction removal can suppress growth if the reaction is essential. Another model change that can suppress growth is to add metabolites to the biomass composition, since this increases the requirements for growth. These two model changes can also be implemented together. However, if neither of these model changes are appropriate, we can suppress growth by removing reactions from the model. To identify the reactions to remove from the model to suppress growth, we implemented an optimization. \n",
    "To illustrate the principles involved in the optimization, we first provide a generic optimization-based solution strategy. We then compare the generic strategy to the optimizations of GrowMatch [[6]](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000308) and Model SEED [[7](https://www.ncbi.nlm.nih.gov/pubmed/19555510),[8](http://www.nature.com/nbt/journal/v28/n9/full/nbt.1672.html)]. GrowMatch, Model SEED, and CROP all solve this problem using a large mixed integer linear program (MILP). \n",
    "All of these MILPs involve the dual linear program to FBA because because the dual characterized by the constraint that growth is not feasible. If these programs tried instead to simply set the FBA growth flux to zero, these programs would only represent that no-growth was feasible; they would not represent that growth is infeasible.\n",
    "\n",
    "## Generic solution strategy ##\n",
    "\n",
    "This strategy independently examines each condition where growth is predicted, but no growth is experimentally observed. For each such condition, a bi-level MILP is created. The purpose of this MILP is to find the optimal set of reactions to remove such that *in silico* growth becomes infeasible in this condition, but remains feasible for a set of control conditions, such as wild-type in minimal medium.\n",
    "This MILP has the same variables as described previously: $S, U, z,$ `weights`, `minimal_growth`, and `atp_maintenance`. However, there are now two conditions: the condition where no growth is experimentally observed, and wild-type in minimal medium. Then, the vector $v$ represents flux for the condition where no growth is experimentally observed, and has upper bounds $U_{no-growth}$; the vector $w$ represents flux for wild-type in minimal medium, and has upper bounds $U_{wt}$.\n",
    "\n",
    "$$\\begin{equation}\\begin{array}{ll}\n",
    "\\min_z  & weights^T(1-z)   \\\\\n",
    "& \n",
    " \\begin{array}{llll}\n",
    "  \\max_v  & v_{biomass} \\\\\n",
    " & Sv=0  &  \\text{ (inner problem}) \\\\\n",
    " &  0\\leq v_i \\leq U_{nogrowth,i}\\cdot z_i  \\\\\n",
    "\\end{array} \\\\\n",
    "& v_{biomass} \\leq  \\text{minimal growth}  \\\\\n",
    "& Sw = 0   \\\\\n",
    "& 0 \\leq w_i \\leq  U_{wt,i}\\cdot z_i  \\\\\n",
    "& w_{ATP} \\geq \\text{atp maintenance} \\\\\n",
    "& z\\in \\{0,1\\}  \\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "The outer minimization problem represents the modeler's optimization to find the optimal set of reactions to remove such that *in silico* growth, represented by $v_{biomass}$, falls below `minimal_growth`. The constraint $v_{biomass}\\leq$ `minimal_growth` implies that when there are no solutions that reduce growth to `minimal_growth`, the MILP will not report on solutions that partially reduce growth. However, such solutions may be helpful in cases where no complete solutions can be identified. \n",
    "The inner maximization problem of the program represents FBA, although without the requirement for ATP. We removed this requirement, because it may not be satisfied in the no-growth condition.\n",
    "The Program is a bi-level MILP. To transform it into a single-level MILP, the inner maximization problem must be transformed into a set of constraints. This has been done using an ingenious technique proposed by Maranas’ lab [[10]](https://www.ncbi.nlm.nih.gov/pubmed/12673766) based on the strong duality theorem of linear programming. In this technique, the inner problem is transformed into a set of constraints by including the constraints of the inner problem’s dual, and then constraining the objectives of the inner problem and the inner problem’s dual to be equal.\n",
    "Before illustrating the primal and dual formulations when $z$ is included, we first show this without $z$ for simplicity. Without $z$, the primal is FBA. To represent FBA in a way that will simplify the derivation of its dual, we represent FBA without unnecessary bounds. So if there is only a single limiting nutrient, e.g. $v_{glucose}\\leq 10$, then we represent FBA with no other upper bounds. Moreover, $v_{glucose}$ does not need an explicit lower bound of zero, because it will never be negative. As with the inner problem above, we also remove the ATP maintenance requirement. Then, the form of the primal LP is:\n",
    "\n",
    "$$\\begin{array}{lll}\n",
    "\\max_v &c^Tv=v_{biomass} \\\\\n",
    "\\text{subject to } &Sv = 0 \\\\\n",
    "& 0\\leq v_i, &\\text{ for $i\\neq$ glucose} \\\\\n",
    "&v_{glucose} \\leq 10\n",
    "\\end{array}$$\n",
    "\n",
    "Now, we need the dual. The reason we need the dual is because we want to model an upper bound on growth. To derive the dual, we follow the approach in [[11]](https://stanford.edu/~boyd/cvxbook/), which uses the  Lagrangian $L$:\n",
    "\n",
    "$$\\begin{array}{rcl}\n",
    "L& = &c^Tv+ mSv - \\sum_{i\\neq glucose} r_i\\cdot v_i+r_{glucose}\\cdot(10 - v_{glucose}) \\\\\n",
    "& = & 10r_{glucose} + (c + S^Tm - r)v \\\\\n",
    "\\end{array}$$\n",
    "\n",
    "Then, the Lagrangian dual problem is:\n",
    "$$\\begin{array}{lll}\n",
    "\\min_{m,r}&\\max_v & 10\\cdot r_{glucose} + (c + S^Tm - r)v \\\\\n",
    "& \\text{subject to}& r_{glucose} \\geq 0 \\\\\n",
    "& & r_i \\leq 0  \\text{   for $i\\neq$ glucose} \\\\\n",
    "\\end{array}$$\n",
    "\n",
    "which is equivalent to:\n",
    "\n",
    "$$\\begin{equation}\\begin{array}{lll}\n",
    "\\min_{m,r}& 10\\cdot r_{glucose}  \\\\\n",
    "\\text{subject to}& S^Tm+c=r \\\\\n",
    "& r_{glucose} \\geq 0 \\\\\n",
    "&r_i \\leq 0 & \\text{for $i\\neq$ glucose} \\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "Where $S^T$ is the transpose of the stochiometric matrix $S$, $r$ is the dual variable associated with reaction bounds (e.g. $v\\geq 0$ and $v_{glucose}\\leq 10$) and $m$ is the dual variable associated with the steady state constraints (that is, for all metabolites, the sum of all the fluxes that produce a metabolite is equal to the sum of the fluxes that consume a metabolite)\n",
    "\n",
    "Now we address the case where there is a binary decision variable per reaction, $z$. To include $z$ in the optimization, we need a loose upper bound, $U$. $U$ is required to be sufficiently large such that the constraint $v_i\\leq U_iz_i$ does not affect the optimum when $z_i=1$. Then, the primal becomes:\n",
    "\n",
    "$$\\begin{equation}\\begin{array}{lll}\n",
    "\\max_v & c^Tv=v_{biomass} \\\\\n",
    "\\text{subject to} & Sv = 0 \\\\\n",
    "& 0\\leq v_i \\leq U_i\\cdot z_i, &\\text{ for all $i$} \\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "where $i$ indexes reactions and $U_{glucose}= 10$.  The dual to this problem is:\n",
    "$$\\begin{equation}\\begin{array}{lll}\n",
    "\\min_r &\\sum_i U_i\\cdot z_i\\cdot r_i \\\\\n",
    "\\text{subject to } &S^Tm+c = r \\\\\n",
    "& r_i\\leq\\Omega_i\\cdot(1-z_i) & \\text{for $i\\neq$ glucose} \\\\\n",
    "& r_{glucose} \\geq 0\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "where $\\Omega_i$ represents a loose upper bound on the elements of the vector $r$, e.g. $\\Omega_i=1000$.  This dual formulation is linear when we treat $z$ as a parameter \"passed in\" from the outer problem to the inner problem. However, the objective is nonlinear in the integrated single-level model, where we include  $z$ as a decision variable along with $r$.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Resolving the nonlinearity in the general solution strategy ###\n",
    "Maranas’s lab has previously shown that they can resolve the nonlinearity in program $(15)$ when $z$ is included as a decision variable [[10]](https://www.ncbi.nlm.nih.gov/pubmed/12673766). However, the solution is highly non-trivial, and we are not aware of any publication that derives it. We infer that they resolved this nonlinearity by exploiting the fact that $U_iz_ir_i=0$ for all reactions $i$ except $i=glucose$. $U_iz_ir_i=0$ for all these reactions because $U_iz_ir_i=0$ when $z_i=0$ and when $z_i=1$. If $z_i=0$, then $U_iz_ir_i=0$ trivially. If $z_i=1$ in the above problem, then the upper bound constraints on $v_i$ will not be tight, since $U_i$ is sufficiently large by design. Constraints that are not tight have dual variables of zero, so if $z_i=1$, then $r_i=0$, implying that $U_iz_ir_i=0$. The reaction $i=glucose$ is slightly different. This reaction is required, so we know it will be included in the model. Thus, we know that $z_{glucose}=1$, so $z_{glucose}$ no longer becomes a decision variable. Furthermore, because glucose is the limiting nutrient, we know that $v_{glucose}=U_{glucose}=10$, so its upper bound $U_{glucose}=10$ is tight. Tightness of this upper bound implies that $r_{glucose}\\geq 0$. \n",
    "By exploiting the behavior of $U_iz_ir_i$, program (15) can be represented as, \n",
    "\n",
    "$$\\begin{equation}\\begin{array}{ll}\n",
    "\\min_{m,r} & 10\\cdot r_{glucose} \\\\\n",
    "\\text{subject to} & S^Tm +c = r \\\\\n",
    "& r_i\\leq\\Omega_i\\cdot(1-z_i)\\ \\text{ for $i\\neq$ glucose} \\\\\n",
    "& r_{glucose} \\geq 0 \\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "Program $(16)$ does not possess the nonlinearity of program $(15)$. Thus program $(16)$ can be used to transform program $(12)$ into a single-level MILP by transforming the inner problem into a set of constraints.\n",
    "\n",
    "$$\\begin{equation}\\begin{array}{l}\n",
    "\\min_z weights^T(1-z) \\\\\n",
    "\\begin{array}{lll}\n",
    "& v_{biomass} = 10r_{glucose} \\\\\n",
    "& Sv= 0 & \\text{inner problem} \\\\\n",
    "& 0\\leq v_i\\leq U_{nogrowth,i}\\cdot z_i \\\\\n",
    "& S^Tm +c = r \\\\\n",
    "& r_i\\leq\\Omega_i\\cdot(1-z_i) & \\text{for $i\\neq$ glucose} \\\\\n",
    "& r_{glucose} \\geq 0 \\\\\n",
    "\\end{array}\\\\\n",
    "v_{biomass} \\leq\\text{minimal growth} \\\\\n",
    "Sw= 0 \\\\\n",
    "0\\leq w_i\\leq U_{wt,i}\\cdot z_i \\\\\n",
    "w_{biomass} \\geq \\text{minimal growth} \\\\\n",
    "w_{ATP} \\geq \\text{atp maintenance} \\\\\n",
    "z\\in \\{0,1\\} \\\\\n",
    "\\end{array}\\end{equation}$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## GrowMatch\n",
    "The main difference in these conditions between the generic solution strategy given in program $(17)$ and GrowMatch [[6](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000308),[12](http://www.genome.org/cgi/doi/10.1101/gr.234503)] concerns how `minimal_growth` is obtained and what its outer problem minimizes. The generic solution strategy constrains $v_{biomass}\\geq$ `minimal_growth`, and its outer problem minimizes a weighted sum of reactions. Instead, GrowMatch’s outer problem minimizes growth, and it constrains the number of gene/reaction removals. This implies that GrowMatch again treats all reactions as having the same weight. In their original paper, they constrained the number of reaction removals per condition to be three [[6]](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000308). In their subsequent paper, they constrained the number of gene/reaction removals per condition to be two [[12]](http://www.genome.org/cgi/doi/10.1101/gr.234503). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## ModelSEED ##\n",
    "The main difference between the generic solution strategy given in program   and the Model SEED [[7](https://www.ncbi.nlm.nih.gov/pubmed/19555510),[8](http://www.nature.com/nbt/journal/v28/n9/full/nbt.1672.html)] is the formulation of the dual in the inner problem. The Model SEED’s dual to FBA has an unexpected form. \n",
    "Putting their dual to FBA into our nomenclature, we get:\n",
    "\n",
    "$$\\begin{equation}\\begin{array}{ll}\n",
    "\\min_{m,r} & 0^Tm + U^Tr \\\\\n",
    "& \\left[\\begin{array}{lr}\n",
    "S & I \\\\\n",
    "0 & -I \\\\\n",
    "0 & I \n",
    "\\end{array}\\right]\\left[\\begin{array}{l}\n",
    "m \\\\\n",
    "r \\\\\n",
    "\\end{array}\\right] \\leq \\left[\\begin{array}{c}\n",
    "K(z-1)+c \\\\\n",
    "-Kz \\\\\n",
    "0 \\end{array}\\right]\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "where $K$ is a large positive constant (e.g. 10,000). Since program $(18)$ is meant as the dual of FBA, we expect *its* dual to be FBA. For comparison, FBA for irreversible reactions is given in program $(14)$. However, when we calculate the dual of this program, we find:\n",
    "\n",
    "$$\\begin{equation}\\begin{array}{ll}\n",
    "\\max_{v,\\tau,\\sigma} & v_{biomass} + Kv(z-1) -K\\tau z \\\\\n",
    "& Sv= 0 \\\\\n",
    "& v-\\tau + \\sigma = U \\\\\n",
    "& v,\\tau,\\sigma \\geq 0 \\\\\n",
    "\\end{array}\\end{equation}$$\n",
    "\n",
    "where $\\tau$ and $\\sigma$ are slack variables, $z$ is treated as a paramter that is \"passed in\" from the outer problem, and $v_{biomass} = c^Tv$. Program (19) is not FBA.\n",
    "\n",
    "This is a problem, because the Model SEED constrains the objective value of program $(18)$ to be equal to the objective value of FBA, whereas the strong duality theorem of linear programming equates the objective value of program $(18)$ and the objective value of program $(19)$. Thus, if the objective value of program $(19)$ differs from the objective value of FBA, then the Model SEED's MILP is infeasible.\n",
    "\n",
    "We refer to program $(19)$ as a softly constrained version of FBA. To see that it is softly constrained, consider the case where only a single reaction $i$ is removed.  This is represented as $z_i= 0$.  In FBA, shown in program $(14)$, we have $0\\leq v_i\\leq U_iz_i$, so $v_i=0$.  However, in this case, program $(19)$ becomes:\n",
    "\n",
    "$$\\begin{array}{ll}\n",
    "\\max_{v,\\tau,\\sigma} & v_{biomass} - K v_i - K\\sum_{j\\neq i}\\tau_j \\\\\n",
    "\\text{subject to} & Sv=0 \\\\\n",
    "& v-\\tau+\\sigma = U \\\\\n",
    "& v,\\tau,\\sigma \\geq 0 \\\\\n",
    "\\end{array}$$\n",
    "\n",
    "Now $v_i$ can be positive if the objective pays a penalty of $Kv_i$. So $z_i=0$  acts as as a soft constraint.  Now consider  a single reaction $j$ with $z_j=1$. Similarly, we see that $v_j$ can exceed its upper bound $U_j$ by $\\tau_j$ if the objective pays a penalty of $K\\tau_j$. So, $U_j$  also acts as a soft constraint. Thus, program $(19)$ treats the constraint $v\\leq Uz$ as a soft constraint in general.\n",
    "\n",
    "To characterize when $v>Uz$, consider the case where increasing the flux through a single reaction $I$ from $vl=U_lZ_l$ to $v_l=U_lz_l+\\Delta v_l$ increases growth by $\\Delta v_{biomass}$. Then, we know that program (19) will penalize the objective by $K\\Delta v_l$, or equivalently if $\\Delta v_{biomass}/\\Delta v_l > K$.  This inequality represents a small additional flux substantially adding to growth.\n",
    "\n",
    "It is unclear how many reactions in the Model SEED's databasecan have $\\Delta v_{biomass}/\\Delta v_l > K$. In our model, we see $\\Delta_{biomass}/\\Delta v_l > 100$, for example with `MYO-IONISITOL-1-PHOSPHATE-SYNTHASE`. This reaction has a large $\\Delta v_{biomass}/\\Delta v_l$ because myo-inositol is needed in small quantities for growth.\n",
    "\n",
    "The possiblity of $\\Delta v_{biomass}/\\Delta v_l > K$ argues for increasing $K$ in program (19).  The problem with increasing $K$, though is that optimization solvers have limited numerical precision, and large numbers in an optimization can have detrimental effects on the precision of the small ones.  However, difficulties with numerical precision of fluxes with high $\\Delta v_{biomass}/\\Delta v_l$ is not limited to the formulation of program (19). In fact, we had trouble ensuring that the constraint $v\\leq Uz$ was enforced strictly because solvers typically allow constraints to be violated by a small margin, such as $10^{-6}$.\n",
    "\n",
    "We emphasize that this discussion does not nullify any of the Model SEED's wonderful results; rather it suggests that fully automated model reconstruction pipelines can still improve.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## CROP ##\n",
    "There are three main differences in these conditions between the generic solution strategy given in program   and CROP. The differences concern CROP’s \n",
    "1. robustness to unbalanced reactions, \n",
    "2. its smaller size, and \n",
    "3. its ability to handle multiple conditions simultaneously. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### ABC model ###"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "![ABC metabolic network](https://raw.githubusercontent.com/pnnl-predictive-phenomics/crop/53507eaab181eecf7c5b3b66a7c5900e87bc87fe/docs/notebook/Metabolic-network.jpeg?token=GHSAT0AAAAAACK5I4RWXSKMK53CD4EU4IKIZTDSASQ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R_1</th>\n",
       "      <th>R_2</th>\n",
       "      <th>R_3</th>\n",
       "      <th>R_4</th>\n",
       "      <th>R_5</th>\n",
       "      <th>R_6</th>\n",
       "      <th>R_7</th>\n",
       "      <th>R_8</th>\n",
       "      <th>R_9</th>\n",
       "      <th>R_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-2.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_ext</th>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C_ext</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D_ext</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E_ext</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_ext</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         R_1    R_2    R_3    R_4    R_5    R_6    R_7    R_8    R_9   R_10\n",
       "A      1.000 -1.000 -1.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
       "B      0.000  1.000  0.000 -1.000  0.000 -2.000  0.000  0.000  0.000  0.000\n",
       "C      0.000  0.000  1.000  0.000  0.000  1.000 -1.000  0.000  0.000 -1.000\n",
       "D      0.000  0.000  0.000  2.000  0.000  0.000  1.000 -1.000  0.000  0.000\n",
       "E      0.000  0.000  0.000 -1.000  1.000  0.000  0.000  0.000  0.000  0.000\n",
       "F      0.000  0.000  0.000  0.000  0.000  1.000  0.000  0.000 -1.000  0.000\n",
       "A_ext -1.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
       "C_ext  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  1.000\n",
       "D_ext  0.000  0.000  0.000  0.000  0.000  0.000  0.000  1.000  0.000  0.000\n",
       "E_ext  0.000  0.000  0.000  0.000 -1.000  0.000  0.000  0.000  0.000  0.000\n",
       "F_ext  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  1.000  0.000"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from cvxpy import *\n",
    "from IPython.display import HTML, Image, display\n",
    "\n",
    "pd.options.display.float_format = \"{:.0f}\".format\n",
    "\n",
    "sys.path.append(\"\")\n",
    "mets = [\"A\", \"B\", \"C\", \"D\", \"E\", \"F\", \"A_ext\", \"C_ext\", \"D_ext\", \"E_ext\", \"F_ext\"]\n",
    "internal_mets = [met for met in mets if \"ext\" not in met]\n",
    "rxns = [\"R_{}\".format(i) for i in range(1, 11)]\n",
    "\n",
    "data = {\n",
    "    \"R_1\": pd.Series({\"A\": 1, \"A_ext\": -1}),\n",
    "    \"R_2\": pd.Series({\"A\": -1, \"B\": 1}),\n",
    "    \"R_3\": pd.Series({\"A\": -1, \"C\": 1}),\n",
    "    \"R_4\": pd.Series({\"B\": -1, \"D\": 2, \"E\": -1}),\n",
    "    \"R_5\": pd.Series({\"E\": 1, \"E_ext\": -1}),\n",
    "    \"R_6\": pd.Series({\"B\": -2, \"C\": 1, \"F\": 1}),\n",
    "    \"R_7\": pd.Series({\"C\": -1, \"D\": 1}),\n",
    "    \"R_8\": pd.Series({\"D\": -1, \"D_ext\": 1}),\n",
    "    \"R_9\": pd.Series({\"F\": -1, \"F_ext\": 1}),\n",
    "    \"R_10\": pd.Series({\"C\": -1, \"C_ext\": 1}),\n",
    "}\n",
    "fullS = pd.DataFrame(data, columns=rxns, index=mets, dtype=\"int64\").fillna(0)\n",
    "\n",
    "pd.options.display.float_format = \"{:.3f}\".format\n",
    "fullS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Robustness ###\n",
    "\n",
    "`CROP` was designed to be robust to unbalanced reactions, because we had unbalanced reactions in our database of metabolic reactions. These unbalanced reactions arose because of difficulties in exporting the MetaCyc reaction to text files, or because a reaction was unbalanced in MetaCyc. Other metabolic reaction databases, such as the Model SEED’s biochemistry database and KEGG, also have unbalanced reactions.\n",
    "Although the existence of such unbalanced reactions is known [13](https://www.ncbi.nlm.nih.gov/pubmed/18697772), we are not aware of any reconstruction algorithm that accounts for them when trying to suppress *in silico* growth. It was not feasible to manually balance all of the unbalanced reactions in our database. Instead, we had to select which of these reactions may belong in our final model. \n",
    "We could not use previous algorithms to select among these unbalanced reactions. To see this, consider the Generic solution strategy. It assumes that the only upper bounds for FBA are those on limiting nutrients, such as glucose. However, an unbalanced reaction could have the form: water $\\rightarrow$ glucose, and therefore act like the limiting nutrient. Because this unbalanced reaction wouldn’t have an upper bound according to the Generic solution strategy, its FBA would be unbounded. (Furthermore, we could not treat these reactions like nutrient sources, because doing so implies that we wouldn’t have any decision variable associated with these reactions, so wouldn’t be able to select which belonged in the final model.)\n",
    "For `CROP` to be robust to unbalanced reactions, we developed a new version of FBA. This version of FBA is called Digital Unlimited-Nutrient Growth FBA (`dung-FBA`). `dung-FBA` is never unbounded, no matter how many unbalanced reactions there are, because it imposes an upper bound on growth: $v_{biomass}\\leq 1$. In fact, this is `dung-FBA`’s only upper bound. `dung-FBA` allows unlimited uptake of the available nutrients, so that the only factor limiting growth is connectivity. This feature is similar to `OnePrune`. Moreover, this allows `dung-FBA` to yield a binary (digital) result to the question: is growth possible? Mathematically, `dung-FBA` is defined by the following linear program:  \n",
    "\n",
    "$$\\begin{array}{lll}\n",
    "\\max_v & c^Tv = v_{biomass} \\\\\n",
    "\\text{subject to} & Sv= 0 &\\text{(dung-FBA)} \\\\\n",
    "&0\\leq v \\\\\n",
    "& v_{biomass} \\leq 1 \\end{array}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### LP dung-FBA implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>v</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          v\n",
       "R_1   0.500\n",
       "R_2   0.500\n",
       "R_3  -0.000\n",
       "R_4   0.500\n",
       "R_5   0.500\n",
       "R_6  -0.000\n",
       "R_7  -0.000\n",
       "R_8   1.000\n",
       "R_9  -0.000\n",
       "R_10 -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "S = fullS.loc[internal_mets].values\n",
    "\n",
    "nmets, nrxns = S.shape\n",
    "v = Variable(nrxns)\n",
    "biomass = 7\n",
    "\n",
    "dung_fba_lp = Problem(Maximize(v[biomass]), [S @ v == 0, 0 <= v, v[biomass] <= 1])\n",
    "dung_fba_lp.solve(solver=cvxpy.SCIPY)\n",
    "display(pd.DataFrame(v.value, index=rxns, columns=[\"v\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "For CROP's MILP, CROP requires the dual of `dung-FBA`. This dual is:\n",
    "\n",
    "$$\\begin{array}{lll}\n",
    "\\min_{m,r} & r_{biomass} \\\\\n",
    "\\text{subject to} & S^Tm + c = r & \\text{(dual of dung-FBA)} \\\\\n",
    "& r_{biomass} \\geq 0 \\\\\n",
    "& r_i\\leq 0 & \\text{for $i\\neq$ biomass} \\end{array}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "####  Dual of LP dung-FBA Implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>r</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          r\n",
       "R_1  -0.000\n",
       "R_2  -0.000\n",
       "R_3  -0.000\n",
       "R_4  -0.000\n",
       "R_5  -0.000\n",
       "R_6  -0.000\n",
       "R_7  -0.000\n",
       "R_8   1.000\n",
       "R_9  -0.000\n",
       "R_10 -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       m\n",
       "A -0.000\n",
       "B -0.000\n",
       "C -0.000\n",
       "D -0.000\n",
       "E -0.000\n",
       "F -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "r = Variable(nrxns)\n",
    "m = Variable(nmets)\n",
    "c = np.zeros(nrxns)\n",
    "c[biomass] = 1\n",
    "dung_fba_dual_lp = Problem(\n",
    "    Minimize(r[biomass]),\n",
    "    [S.T @ m + c == r, r[:biomass] <= 0, r[biomass] >= 0, r[biomass + 1 :] <= 0],\n",
    ")\n",
    "dung_fba_dual_lp.solve(solver=cvxpy.SCIPY)\n",
    "display(pd.DataFrame(r.value, index=rxns, columns=[\"r\"]))\n",
    "display(pd.DataFrame(m.value, index=internal_mets, columns=[\"m\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "where $r$ is the dual variable associated with reaction constraints, and $m$ is the dual variable associated with metabolite constraints.  Now we add binary decision variables, $z$, so that we can select reactions to remove. `dung-FBA` now becomes\n",
    "\n",
    "$$\\begin{array}{lll}\n",
    "\\max_{v} & c^Tv = v_{biomass} \\\\\n",
    "& Sv = 0 \\\\\n",
    "& 0\\leq v_i \\leq U_i z_i & \\text{for all $i$}\n",
    "\\end{array}$$\n",
    "\n",
    "where $U_{biomass}=1$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### MILP dung-FBA Implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>v</th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>0.500</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>0.500</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>-0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>0.500</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>0.500</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>-0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>-0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>-0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          v     z\n",
       "R_1   0.500 1.000\n",
       "R_2   0.500 1.000\n",
       "R_3  -0.000 1.000\n",
       "R_4   0.500 1.000\n",
       "R_5   0.500 1.000\n",
       "R_6  -0.000 1.000\n",
       "R_7  -0.000 1.000\n",
       "R_8   1.000 1.000\n",
       "R_9   0.000 1.000\n",
       "R_10 -0.000 1.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "U = np.eye(nrxns) * 1000\n",
    "U[biomass, biomass] = 1\n",
    "z = Variable(nrxns, boolean=True)\n",
    "dung_fba_milp = Problem(Maximize(v[biomass]), [S @ v == 0, 0 <= v, v <= U @ z])\n",
    "dung_fba_milp.solve(solver=cvxpy.SCIPY)\n",
    "reactions = pd.DataFrame(np.squeeze(np.asarray(v.value)), index=rxns, columns=[\"v\"])\n",
    "reactions[\"z\"] = pd.Series(np.squeeze(np.asarray(z.value)), index=rxns)\n",
    "display(reactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The dual of this program is:\n",
    "$$\\begin{array}{lll}\n",
    "\\min_{m,r} & r_{biomass} \\\\\n",
    "\\text{subject to} & S^Tm + c = r \\\\\n",
    "& r_{biomass} \\geq 0 \\\\\n",
    "& r_i\\leq \\Omega_i\\cdot(1-z_i) & \\text{for $i\\neq$ biomass} \\end{array}$$\n",
    "\n",
    "For reactions $i$ that are knocked out as a result of a gene knockout, $z_i$ is treated as 0, implying that $r_i\\leq\\Omega_i$. Since $\\Omega_i$ is supposed to be a loose upper bound, we represented this as $r_i$ being free.\n",
    "\n",
    "We used `dung-FBA` along with `limed-FBA` by replacing the $S$ in `dung-FBA` with `limed-FBA`'s $S_{limed}$.  Using dung-FBA successfully brought us robustness to unbalanced reactions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Dual of MILP dung-FBA  Implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>0.000</td>\n",
       "      <td>-2998.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>0.000</td>\n",
       "      <td>-3000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>0.000</td>\n",
       "      <td>-1999.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>0.000</td>\n",
       "      <td>-2000.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z         r\n",
       "R_1  0.000  1000.000\n",
       "R_2  0.000  1000.000\n",
       "R_3  0.000  1000.000\n",
       "R_4  0.000 -2998.000\n",
       "R_5  0.000  1000.000\n",
       "R_6  0.000 -3000.000\n",
       "R_7  0.000 -1999.000\n",
       "R_8  0.000     0.000\n",
       "R_9  0.000  1000.000\n",
       "R_10 0.000 -2000.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>2000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>2000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-1000.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          m\n",
       "A  1000.000\n",
       "B  2000.000\n",
       "C  2000.000\n",
       "D     1.000\n",
       "E  1000.000\n",
       "F -1000.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Omega = 1000\n",
    "\n",
    "m = Variable(nmets)\n",
    "r = Variable(nrxns)\n",
    "z = Variable(nrxns, boolean=True)\n",
    "\n",
    "dung_fba_dual_milp = Problem(\n",
    "    Minimize(r[biomass]),\n",
    "    [\n",
    "        S.T @ m + c == r,\n",
    "        r[:biomass] <= Omega * np.eye(biomass) @ (1 - z[:biomass]),\n",
    "        r[biomass] >= 0,\n",
    "        r[biomass + 1 :] <= Omega * np.eye(nrxns - biomass - 1) @ (1 - z[biomass + 1 :]),\n",
    "    ],\n",
    ")\n",
    "dung_fba_dual_milp.solve(solver=cvxpy.SCIPY)\n",
    "\n",
    "\n",
    "reactions = pd.DataFrame(z.value, index=rxns, columns=[\"z\"])\n",
    "reactions[\"r\"] = pd.Series(np.squeeze(np.asarray(r.value)), index=rxns)\n",
    "display(reactions)\n",
    "metabolites = pd.DataFrame(np.squeeze(np.asarray(m.value)), index=internal_mets, columns=[\"m\"])\n",
    "display(metabolites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Size ###\n",
    "\n",
    "FBA maximizes growth, so its dual represents an upper bound on growth. Thus, to mathematically express the experimental observation of no-growth (or of growth below a threshold), we need the dual. However, it also follows that once we have the dual, we no longer need FBA. In fact, incorporating both FBA and its dual only provides value if we want to model an intermediate level of growth. This observation appears to have been overlooked. However, only using the dual substantially reduces the MILP’s size.\n",
    "This observation is actually suggested by program  (17) in *Generic solution strategy*. Program  (17) includes the constraints $v_{biomass}=10r_{glucose}$ and it includes $v_{biomass}\\leq$ `minimal_growth`. Together, these constraints imply that $10r_{glucose}\\leq$ `minimal_growth`, and that we can ignore $v_{biomass}$. When we remove FBA’s growth maximization from the inner problem of program (17), we get:\n",
    "\n",
    "$$\\begin{array}{ll}\n",
    "\\min_z &  weights^T(1-z) \\\\\n",
    " &\\begin{array}{lll}\n",
    "& S^Tm -r = -c \\\\\n",
    "& r_i\\leq\\Omega_i\\cdot(1-z_i) & \\text{for $i\\neq$ glucose} \\\\\n",
    "& r_{glucose} \\geq 0 \\end{array} \\\\\n",
    "& 10r_{glucose} \\leq \\text{minimal growth} \\\\\n",
    "& Sw = 0 \\\\\n",
    "& 0\\leq w_i \\leq U_{wt,i}\\cdot z_i \\\\\n",
    "& w_{biomass} \\geq \\text{minimal growth} \\\\\n",
    "& w_{ATP} \\geq \\text{atp maintenance} \\\\\n",
    "& z\\in \\{0,1\\}\n",
    "\\end{array}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Reduced bi-level MILP dung-FBA implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        z     r     w\n",
       "R_1   NaN  None  None\n",
       "R_2   NaN  None  None\n",
       "R_3   NaN  None  None\n",
       "R_4   NaN  None  None\n",
       "R_5   NaN  None  None\n",
       "R_6   NaN  None  None\n",
       "R_7   NaN  None  None\n",
       "R_8   NaN  None  None\n",
       "R_9   NaN  None  None\n",
       "R_10  NaN  None  None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     m\n",
       "A  NaN\n",
       "B  NaN\n",
       "C  NaN\n",
       "D  NaN\n",
       "E  NaN\n",
       "F  NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glucose = 4\n",
    "ATP = 8\n",
    "minimal_growth = 0.1\n",
    "atp_maintenance = 0.1\n",
    "\n",
    "U = np.eye(nrxns) * 1000.0\n",
    "weights = np.ones(nrxns) * 0.5\n",
    "Omega = 1000.0\n",
    "\n",
    "\n",
    "w = Variable(nrxns)\n",
    "m = Variable(nmets)\n",
    "r = Variable(nrxns)\n",
    "z = Variable(nrxns, boolean=True)\n",
    "\n",
    "dung_fba_reduced = Problem(\n",
    "    Minimize(weights.T @ (1 - z)),\n",
    "    [\n",
    "        S.T @ m - r == -c,\n",
    "        r[:glucose] <= Omega * np.eye(glucose) @ (1 - z[:glucose]),\n",
    "        r[glucose] >= 0,\n",
    "        r[glucose + 1 :] <= Omega * np.eye(nrxns - glucose - 1) @ (1 - z[glucose + 1 :]),\n",
    "        10 * r[glucose] <= minimal_growth,\n",
    "        S @ w == 0,\n",
    "        0 <= w,\n",
    "        w <= U @ z,\n",
    "        w[biomass] >= minimal_growth,\n",
    "        w[ATP] >= atp_maintenance,\n",
    "    ],\n",
    ")\n",
    "dung_fba_reduced.solve()\n",
    "reactions = pd.DataFrame(z.value, index=rxns, columns=[\"z\"])\n",
    "reactions[\"r\"] = pd.Series(np.squeeze(np.asarray(r.value)), index=rxns)\n",
    "reactions[\"w\"] = pd.Series(np.squeeze(np.asarray(r.value)), index=rxns)\n",
    "display(reactions)\n",
    "metabolites = pd.DataFrame(m.value, index=internal_mets, columns=[\"m\"])\n",
    "display(metabolites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "One issue with the Generic solution strategy is that when there are no solutions that reduce growth to `minimal_growth`, solutions that partially reduce growth are not reported. To overcome this issue, $r_{glucose}$ can be moved to the objective.\n",
    "\n",
    "$$\\begin{array}{ll}\n",
    "\\min_z &  weights^T(1 - z)  + B\\cdot r_{glucose}\\\\\n",
    " &\\begin{array}{lll}\n",
    "& S^Tm -r = -c \\\\\n",
    "& r_i\\leq\\Omega_i\\cdot(1-z_i) & \\text{for $i\\neq$ glucose} \\\\\n",
    "& r_{glucose} \\geq 0 \\end{array} \\\\\n",
    "& Sw = 0 \\\\\n",
    "& 0\\leq w_i \\leq U_{wt,i}\\cdot z_i \\\\\n",
    "& w_{biomass} \\geq \\text{minimal growth} \\\\\n",
    "& w_{ATP} \\geq \\text{atp maintenance} \\\\\n",
    "& z\\in \\{0,1\\}\n",
    "\\end{array}$$\n",
    "\n",
    "where the constant $B$ balances\n",
    "1. the model's consistency with reaction weights versus\n",
    "2. consistency with growth phenotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Partially reduced bi-level MILP dung-FBA implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>1.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>1.000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z      r      w\n",
       "R_1  1.000  0.000  0.250\n",
       "R_2  1.000  0.000  0.250\n",
       "R_3  1.000  0.000 -0.000\n",
       "R_4  1.000  0.000  0.050\n",
       "R_5  1.000  2.000  0.050\n",
       "R_6  1.000 -0.000  0.100\n",
       "R_7  0.000  1.000 -0.000\n",
       "R_8  1.000  0.000  0.100\n",
       "R_9  1.000  0.000  0.100\n",
       "R_10 1.000  0.000  0.100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>2.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       m\n",
       "A  0.000\n",
       "B  0.000\n",
       "C -0.000\n",
       "D  1.000\n",
       "E  2.000\n",
       "F -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glucose = 4\n",
    "ATP = 8\n",
    "\n",
    "minimal_growth = 0.1\n",
    "atp_maintenance = 0.1\n",
    "\n",
    "B = 1000\n",
    "U = np.eye(nrxns) * 1000.0\n",
    "weights = np.ones(nrxns) * 0.5\n",
    "Omega = 1000.0\n",
    "\n",
    "w = Variable(nrxns)\n",
    "m = Variable(nmets)\n",
    "r = Variable(nrxns)\n",
    "z = Variable(nrxns, boolean=True)\n",
    "\n",
    "dung_fba_partially_reduced = Problem(\n",
    "    Minimize(weights.T @ (1 - z) + B * r[glucose]),\n",
    "    [\n",
    "        S.T @ m - r == -c,\n",
    "        r[:glucose] <= Omega * np.eye(glucose) @ (1 - z[:glucose]),\n",
    "        r[glucose] >= 0,\n",
    "        r[glucose + 1 :] <= Omega * np.eye(nrxns - glucose - 1) @ (1 - z[glucose + 1 :]),\n",
    "        S @ w == 0,\n",
    "        0 <= w,\n",
    "        w <= U @ z,\n",
    "        w[biomass] >= minimal_growth,\n",
    "        w[ATP] >= atp_maintenance,\n",
    "    ],\n",
    ")\n",
    "dung_fba_partially_reduced.solve(solver=cvxpy.SCIPY)\n",
    "reactions = pd.DataFrame(z.value, index=rxns, columns=[\"z\"])\n",
    "reactions[\"r\"] = pd.Series(np.squeeze(np.asarray(r.value)), index=rxns)\n",
    "reactions[\"w\"] = pd.Series(np.squeeze(np.asarray(w.value)), index=rxns)\n",
    "display(reactions)\n",
    "metabolites = pd.DataFrame(m.value, index=internal_mets, columns=[\"m\"])\n",
    "display(metabolites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Now, we replace the dual of FBA with the dual of `dung-FBA`, which involves substituting $r_{glucose}$ with $r_{biomass}$.\n",
    "\n",
    "$$\\begin{array}{ll}\n",
    "\\min_z &  weights^T(1-z)  + B\\cdot r_{biomass}\\\\\n",
    " &\\begin{array}{llll}\n",
    "& S^Tm -r = -c \\\\\n",
    "& r_i\\leq\\Omega_i\\cdot(1-z_i) & \\text{for $i\\neq$ biomass} & \\text{(dual of dung-FBA)}\\\\\n",
    "& r_{biomass} \\geq 0 \\end{array} \\\\\n",
    "& Sw = 0 \\\\\n",
    "& 0\\leq w_i \\leq U_{wt,i}\\cdot z_i & \\text{(FBA)} \\\\\n",
    "& w_{biomass} \\geq \\text{minimal growth} \\\\\n",
    "& w_{ATP} \\geq \\text{atp maintenance} \\\\\n",
    "& z\\in \\{0,1\\}\n",
    "\\end{array}$$\n",
    "\n",
    "A numerical difficulty that arises is that although $z_i$ is supposed to be binary, its value may differ from 0 or 1 by a small amount, such as by $10^{-6}$. Thus, even when $z_i$ is supposed to be 1, $r_i$ may not be forced to be negative, but may instead reach $10^{-6}\\cdot \\Omega_i$. To mitigate the effects of such numerical difficulties, we had some success with using rather small $\\Omega_i$ for all reactions, e.g. $\\Omega_i=1$. We found that using such tight upper bounds also improved run-time, which is consistent with the findings of others [[14]](https://www.ncbi.nlm.nih.gov/pubmed/20426856).\n",
    "\n",
    "This section shows that `CROP` deals with conditions where growth is predicted but no growth is experimentally observed with a MILP that is much smaller than previous algorithms, allowing for more efficient solution."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Biomass-substituted partially reduced bi-level MILP dung-FBA implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Omega = 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>1.000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z      r      w\n",
       "R_1  1.000  0.000  0.200\n",
       "R_2  1.000  0.000  0.200\n",
       "R_3  1.000  0.000 -0.000\n",
       "R_4  1.000 -0.000 -0.000\n",
       "R_5  1.000  0.000  0.000\n",
       "R_6  1.000  0.000  0.100\n",
       "R_7  1.000 -0.000  0.100\n",
       "R_8  1.000  1.000  0.100\n",
       "R_9  1.000  0.000  0.100\n",
       "R_10 1.000  0.000 -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       m\n",
       "A  0.000\n",
       "B  0.000\n",
       "C  0.000\n",
       "D  0.000\n",
       "E -0.000\n",
       "F -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Omega = 100\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z     r      w\n",
       "R_1  1.000 0.000  0.250\n",
       "R_2  1.000 0.000  0.250\n",
       "R_3  1.000 0.000 -0.000\n",
       "R_4  1.000 0.000  0.050\n",
       "R_5  1.000 0.000  0.050\n",
       "R_6  1.000 0.000  0.100\n",
       "R_7  1.000 0.000 -0.000\n",
       "R_8  1.000 1.000  0.100\n",
       "R_9  1.000 0.000  0.100\n",
       "R_10 1.000 0.000  0.100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       m\n",
       "A  0.000\n",
       "B  0.000\n",
       "C  0.000\n",
       "D  0.000\n",
       "E  0.000\n",
       "F -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Omega = 1000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z     r      w\n",
       "R_1  1.000 0.000  0.250\n",
       "R_2  1.000 0.000  0.250\n",
       "R_3  1.000 0.000 -0.000\n",
       "R_4  1.000 0.000  0.050\n",
       "R_5  1.000 0.000  0.050\n",
       "R_6  1.000 0.000  0.100\n",
       "R_7  1.000 0.000 -0.000\n",
       "R_8  1.000 1.000  0.100\n",
       "R_9  1.000 0.000  0.100\n",
       "R_10 1.000 0.000  0.100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       m\n",
       "A  0.000\n",
       "B  0.000\n",
       "C  0.000\n",
       "D  0.000\n",
       "E  0.000\n",
       "F -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Omega = 10000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z     r      w\n",
       "R_1  1.000 0.000  0.250\n",
       "R_2  1.000 0.000  0.250\n",
       "R_3  1.000 0.000 -0.000\n",
       "R_4  1.000 0.000  0.050\n",
       "R_5  1.000 0.000  0.050\n",
       "R_6  1.000 0.000  0.100\n",
       "R_7  1.000 0.000 -0.000\n",
       "R_8  1.000 1.000  0.100\n",
       "R_9  1.000 0.000  0.100\n",
       "R_10 1.000 0.000  0.100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       m\n",
       "A  0.000\n",
       "B  0.000\n",
       "C  0.000\n",
       "D  0.000\n",
       "E  0.000\n",
       "F -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glucose = 4\n",
    "ATP = 8\n",
    "minimal_growth = 0.1\n",
    "atp_maintenance = 0.1\n",
    "B = 1000\n",
    "\n",
    "U = np.eye(nrxns) * 1000.0\n",
    "weights = np.ones(nrxns) * 0.5\n",
    "\n",
    "w = Variable(nrxns)\n",
    "m = Variable(nmets)\n",
    "r = Variable(nrxns)\n",
    "z = Variable(nrxns, boolean=True)\n",
    "\n",
    "for Omega in [1, 100, 1000, 10000]:\n",
    "    dung_fba_partially_reduced_biomass = Problem(\n",
    "        Minimize(weights.T @ (1 - z) + B * r[biomass]),\n",
    "        [\n",
    "            S.T @ m - r == -c,\n",
    "            r[:biomass] <= Omega * np.eye(biomass) @ (1 - z[:biomass]),\n",
    "            r[biomass] >= 0,\n",
    "            r[biomass + 1 :] <= Omega * np.eye(nrxns - biomass - 1) @ (1 - z[biomass + 1 :]),\n",
    "            S @ w == 0,\n",
    "            0 <= w,\n",
    "            w <= U @ z,\n",
    "            w[biomass] >= minimal_growth,\n",
    "            w[ATP] >= atp_maintenance,\n",
    "        ],\n",
    "    )\n",
    "    dung_fba_partially_reduced_biomass.solve(solver=cvxpy.SCIPY)\n",
    "    reactions = pd.DataFrame(z.value, index=rxns, columns=[\"z\"])\n",
    "    reactions[\"r\"] = pd.Series(np.squeeze(np.asarray(r.value)), index=rxns)\n",
    "    reactions[\"w\"] = pd.Series(np.squeeze(np.asarray(w.value)), index=rxns)\n",
    "    print(\"Omega = {}\".format(Omega))\n",
    "    display(reactions)\n",
    "    metabolites = pd.DataFrame(m.value, index=internal_mets, columns=[\"m\"])\n",
    "    display(metabolites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Handling multiple conditions simultaneously ###\n",
    "\n",
    "Facilitated by its robustness and smaller size, `CROP` can simultaneously handle multiple conditions where no growth is experimentally observed. For two conditions where no-growth is experimentally observed, the MILP is similar to above, albeit with an additional dual.\n",
    "\n",
    "$$\\begin{array}{ll}\n",
    "\\min_z &  weights^T(1-z)  + B\\cdot( r_{1,biomass} + r_{2,biomass})\\\\\n",
    " &\\begin{array}{lll}\n",
    "& S^Tm_1 -r_1 = -c \\\\\n",
    "& r_{1,biomass} \\geq 0 &  \\text{(dual 1)}\\\\\n",
    "& r_{1,i}\\leq\\Omega_{1,i}\\cdot(1-z_i) & \\text{for $i\\neq$ biomass} \n",
    "\\end{array} \\\\\n",
    "& \\\\\n",
    " &\\begin{array}{lll}\n",
    "& S^Tm_2 -r_2 = -c \\\\\n",
    "& r_{2,biomass} \\geq 0 &  \\text{(dual 2)}\\\\\n",
    "& r_{2,i}\\leq\\Omega_{2,i}\\cdot(1-z_i) & \\text{for $i\\neq$ biomass} \n",
    "\\end{array} \\\\\n",
    "& Sw = 0  \\\\\n",
    "& 0\\leq w_i \\leq U_{wt,i}\\cdot z_i & \\text{(FBA)} \\\\\n",
    "& w_{biomass} \\geq \\text{minimal growth} \\\\\n",
    "& w_{ATP} \\geq \\text{atp maintenance} \\\\\n",
    "& z\\in \\{0,1\\}\n",
    "\\end{array}$$\n",
    "\n",
    "Handling multiple conditions simultaneously allows for identification of model modifications that synergistically solve these conditions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Multiple condition biomass-substituted partially reduced bi-level MILP dung-FBA Implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r1</th>\n",
       "      <th>r2</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z    r1    r2      w\n",
       "R_1  1.000 0.000 0.000  0.200\n",
       "R_2  1.000 0.000 0.000  0.200\n",
       "R_3  1.000 0.000 0.000 -0.000\n",
       "R_4  1.000 0.000 0.000 -0.000\n",
       "R_5  1.000 0.000 0.000  0.000\n",
       "R_6  1.000 0.000 0.000  0.100\n",
       "R_7  1.000 0.000 0.000  0.100\n",
       "R_8  1.000 1.000 1.000  0.100\n",
       "R_9  1.000 0.000 0.000  0.100\n",
       "R_10 1.000 0.000 0.000 -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m1</th>\n",
       "      <th>m2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     m1     m2\n",
       "A 0.000  0.000\n",
       "B 0.000 -0.000\n",
       "C 0.000  0.000\n",
       "D 0.000  0.000\n",
       "E 0.000  0.000\n",
       "F 0.000 -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glucose = 0\n",
    "ATP = 8\n",
    "minimal_growth = 0.1\n",
    "atp_maintenance = 0.1\n",
    "B = 10\n",
    "\n",
    "U = np.eye(nrxns) * 1000.0\n",
    "weights = np.ones(nrxns) * 0.5\n",
    "\n",
    "Omega_before = np.eye(biomass)\n",
    "\n",
    "Omega_after = np.eye(nrxns - biomass - 1)\n",
    "\n",
    "w = cvxpy.Variable(nrxns)\n",
    "m1 = cvxpy.Variable(nmets)\n",
    "m2 = cvxpy.Variable(nmets)\n",
    "\n",
    "r1 = cvxpy.Variable(nrxns)\n",
    "r2 = cvxpy.Variable(nrxns)\n",
    "z = cvxpy.Variable(nrxns, boolean=True)\n",
    "\n",
    "dung_fba_multiple_conditions = cvxpy.Problem(\n",
    "    cvxpy.Minimize(weights.T @ (1 - z) + B * (r1[biomass] + r2[biomass])),\n",
    "    constraints=[\n",
    "        S.T @ m1 - r1 == -c,\n",
    "        r1[:biomass] == Omega_before @ (1 - z[:biomass]),\n",
    "        r1[biomass] >= 0,\n",
    "        r1[biomass + 1 :] >= Omega_after @ (1 - z[biomass + 1 :]),\n",
    "        S.T @ m2 - r2 == -c,\n",
    "        r2[:biomass] == np.eye(biomass) @ (1 - z[:biomass]),\n",
    "        r2[biomass] >= 0,\n",
    "        r2[biomass + 1 :] >= Omega_after @ (1 - z[biomass + 1 :]),\n",
    "        S @ w == 0,\n",
    "        0 <= w,\n",
    "        w <= U @ z,\n",
    "        w[biomass] >= minimal_growth,\n",
    "        w[ATP] >= atp_maintenance,\n",
    "    ],\n",
    ")\n",
    "\n",
    "dung_fba_multiple_conditions.solve(\n",
    "    solver=cvxpy.SCIPY,\n",
    ")\n",
    "reactions = pd.DataFrame(z.value, index=rxns, columns=[\"z\"])\n",
    "reactions[\"r1\"] = pd.Series(np.squeeze(np.asarray(r1.value)), index=rxns)\n",
    "reactions[\"r2\"] = pd.Series(np.squeeze(np.asarray(r2.value)), index=rxns)\n",
    "reactions[\"w\"] = pd.Series(np.squeeze(np.asarray(w.value)), index=rxns)\n",
    "display(reactions)\n",
    "metabolites = pd.DataFrame(np.squeeze(np.asarray(m1.value)), index=internal_mets, columns=[\"m1\"])\n",
    "metabolites[\"m2\"] = pd.Series(np.squeeze(np.asarray(m2.value)), index=internal_mets)\n",
    "display(metabolites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Multiple condition nutrient-limited partially reduced bi-level MILP dung-FBA Implementation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1., 0., 0., 0., 0.],\n",
       "       [0., 1., 0., 0., 0.],\n",
       "       [0., 0., 1., 0., 0.],\n",
       "       [0., 0., 0., 1., 0.],\n",
       "       [0., 0., 0., 0., 1.]])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.eye(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>r1</th>\n",
       "      <th>r2</th>\n",
       "      <th>w</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R_1</th>\n",
       "      <td>1.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_3</th>\n",
       "      <td>1.000</td>\n",
       "      <td>-2.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_5</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_6</th>\n",
       "      <td>1.000</td>\n",
       "      <td>-4.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_7</th>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_8</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_9</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_10</th>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         z     r1    r2      w\n",
       "R_1  1.000  2.000 0.000  0.250\n",
       "R_2  1.000  0.000 0.000  0.250\n",
       "R_3  1.000 -2.000 0.000 -0.000\n",
       "R_4  1.000  0.000 0.000  0.050\n",
       "R_5  1.000  0.000 2.000  0.050\n",
       "R_6  1.000 -4.000 0.000  0.100\n",
       "R_7  0.000  1.000 1.000 -0.000\n",
       "R_8  1.000  0.000 0.000  0.100\n",
       "R_9  1.000  0.000 0.000  0.100\n",
       "R_10 1.000  0.000 0.000  0.100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m1</th>\n",
       "      <th>m2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>2.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>2.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000</td>\n",
       "      <td>2.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.000</td>\n",
       "      <td>-0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      m1     m2\n",
       "A  2.000  0.000\n",
       "B  2.000  0.000\n",
       "C  0.000  0.000\n",
       "D  1.000  1.000\n",
       "E  0.000  2.000\n",
       "F -0.000 -0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nutrient1 = 0\n",
    "nutrient2 = 4\n",
    "\n",
    "ATP = 8\n",
    "minimal_growth = 0.1\n",
    "atp_maintenance = 0.1\n",
    "B = 10\n",
    "\n",
    "U = np.eye(nrxns) * 1000.0\n",
    "weights = np.ones(nrxns) * 0.5\n",
    "Omega1 = 1000\n",
    "Omega2 = 1000\n",
    "\n",
    "w = Variable(nrxns)\n",
    "m1 = Variable(nmets)\n",
    "m2 = Variable(nmets)\n",
    "\n",
    "r1 = Variable(nrxns)\n",
    "r2 = Variable(nrxns)\n",
    "z = Variable(nrxns, boolean=True)\n",
    "\n",
    "dung_fba_multiple_conditions = Problem(\n",
    "    Minimize(weights.T @ (1 - z) + B * (r1[nutrient1] + r2[nutrient2])),\n",
    "    [\n",
    "        S.T @ m1 - r1 == -c,\n",
    "        # r1[:nutrient1]      <= Omega1*np.eye(nutrient1)*(1 - z[:nutrient1]),\n",
    "        r1[nutrient1] >= 0,\n",
    "        r1[nutrient1 + 1 :] <= Omega1 * np.eye(nrxns - nutrient1 - 1) @ (1 - z[nutrient1 + 1 :]),\n",
    "        S.T @ m2 - r2 == -c,\n",
    "        r2[:nutrient2] <= Omega2 * np.eye(nutrient2) @ (1 - z[:nutrient2]),\n",
    "        r2[nutrient2] >= 0,\n",
    "        r2[nutrient2 + 1 :] <= Omega2 * np.eye(nrxns - nutrient2 - 1) @ (1 - z[nutrient2 + 1 :]),\n",
    "        S @ w == 0,\n",
    "        0 <= w,\n",
    "        w <= U @ z,\n",
    "        w[biomass] >= minimal_growth,\n",
    "        w[ATP] >= atp_maintenance,\n",
    "    ],\n",
    ")\n",
    "dung_fba_multiple_conditions.solve(solver=cvxpy.SCIPY)\n",
    "reactions = pd.DataFrame(z.value, index=rxns, columns=[\"z\"])\n",
    "reactions[\"r1\"] = pd.Series(np.squeeze(np.asarray(r1.value)), index=rxns)\n",
    "reactions[\"r2\"] = pd.Series(np.squeeze(np.asarray(r2.value)), index=rxns)\n",
    "reactions[\"w\"] = pd.Series(np.squeeze(np.asarray(w.value)), index=rxns)\n",
    "display(reactions)\n",
    "metabolites = pd.DataFrame(np.squeeze(np.asarray(m1.value)), index=internal_mets, columns=[\"m1\"])\n",
    "metabolites[\"m2\"] = pd.Series(np.squeeze(np.asarray(m2.value)), index=internal_mets)\n",
    "display(metabolites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# References #\n",
    "1. Arakaki A, Huang Y, Skolnick J (2009) EFICAz2: enzyme function inference by a combined approach enhanced by machine learning. BMC Bioinformatics 10: 107.\n",
    "2. Jankowski M, Henry C, Broadbelt L, Hatzimanikatis V (2008) Group Contribution Method for Thermodynamic Analysis of Complex Metabolic Networks. Biophysical Journal 95: 1487-1499.\n",
    "3. Fleming RM, Thiele I, Nasheuer HP (2009) Quantitative assignment of reaction directionality in constraint-based models of metabolism: Application to Escherichia coli. Biophysical Chemistry 145: 47-56.\n",
    "4. Bar-Even A, Noor E, Flamholz A, Buescher J, Milo R (2011) Hydrophobicity and charge shape cellular metabolite concentrations. PLoS computational biology 7: e1002166.\n",
    "5. Henry C, Jankowski M, Broadbelt L, Hatzimanikatis V (2006) Genome-Scale Thermodynamic Analysis of Escherichia coli Metabolism. Biophysical Journal 90: 1453-1461.\n",
    "6. Kumar V, Maranas C (2009) GrowMatch: An Automated Method for Reconciling In Silico/In Vivo Growth Predictions. PLoS Comput Biol 5: e1000308.\n",
    "7. Henry C, Zinner J, Cohoon M, Stevens R (2009) iBsu1103: a new genome-scale metabolic model of Bacillus subtilis based on SEED annotations. Genome Biology 10: R69.\n",
    "8. Henry C, DeJongh M, Best A, Frybarger P, Linsay B, et al. (2010) High-throughput generation, optimization and analysis of genome-scale metabolic models. Nat Biotech 28: 977-982.\n",
    "9. Mitchell J (2001) Restarting after Branching in the SDP Approach to MAX-CUT and Similar Combinatorial Optimization Problems http://www.nature.com/nbt/journal/v28/n9/full/nbt.1672.html. Journal of Combinatorial Optimization 5: 151-166.\n",
    "10. Burgard A, Maranas C (2003) [Optimization-based framework for inferring and testing hypothesized metabolic objective functions](https://www.ncbi.nlm.nih.gov/pubmed/12673766). Biotechnology and bioengineering 82: 670-677.\n",
    "11. Boyd S, Vandenberghe L (2004) Convex Optimization: Cambridge University Press.\n",
    "12. Zomorrodi A, Maranas C (2010) Improving the iMM904 S. cerevisiae metabolic model using essentiality and synthetic lethality data. BMC Systems Biology 4: 178.\n",
    "13. Gevorgyan A, Poolman M, Fell D (2008) Detection of stoichiometric inconsistencies in biomolecular models. Bioinformatics 24: 2245-2251.\n",
    "14. Kim J, Reed J (2010) OptORF: Optimal metabolic and regulatory perturbations for metabolic engineering of microbial strains. BMC Systems Biology 4: 53.\n"
   ]
  }
 ],
 "metadata": {
  "_draft": {
   "nbviewer_url": "https://gist.github.com/8f988e25afa3854df6602473e8298cb9"
  },
  "gist": {
   "data": {
    "description": "Consistent Reproduction Of Phenotype working implementation",
    "public": true
   },
   "id": "8f988e25afa3854df6602473e8298cb9"
  },
  "kernelspec": {
   "display_name": "CROP_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  },
  "latex_envs": {
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 0
  },
  "toc": {
   "toc_cell": true,
   "toc_number_sections": true,
   "toc_section_display": "block",
   "toc_threshold": 6,
   "toc_window_display": true
  },
  "toc_position": {
   "height": "296px",
   "left": "1068.88px",
   "right": "20px",
   "top": "120px",
   "width": "300px"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
